src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8.cs

```diff
@@ -590,6 +606,53 @@ static bool GrowAndAppendFormatted(scoped ref TryWriteInterpolatedStringHandler
                 }
             }
 
+            // TODO https://github.com/dotnet/runtime/issues/81500:
+            // Remove once Enum.TryFormat(Span<byte>, ...) is available.
+            /// <summary>Writes the specified enum to the handler.</summary>
+            /// <param name="value">The value to write. It must be an enum but isn't constrained because the caller doesn't have a constraint.</param>
+            /// <param name="format">The format string.</param>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private bool AppendEnum<T>(T value, string? format)
+            {
+                Debug.Assert(typeof(T).IsEnum);
+
+                Span<char> utf16 = stackalloc char[256];
+                return Enum.TryFormatUnconstrained(value, utf16, out int charsWritten, format) ?
+                    AppendFormatted(utf16.Slice(0, charsWritten)) :
+                    GrowAndAppendFormatted(ref this, value, utf16.Length, out charsWritten, format);
+
+                [MethodImpl(MethodImplOptions.NoInlining)]
+                static bool GrowAndAppendFormatted(scoped ref TryWriteInterpolatedStringHandler thisRef, T value, int length, out int charsWritten, string? format)
+                {
+                    Debug.Assert(value is ISpanFormattable);
+
+                    while (true)
+                    {
+                        int newLength = length * 2;
+                        if ((uint)newLength > Array.MaxLength)
+                        {
+                            newLength = length == Array.MaxLength ?
+                                Array.MaxLength + 1 : // force OOM
+                                Array.MaxLength;
+                        }
+                        length = newLength;
+
+                        char[] array = ArrayPool<char>.Shared.Rent(length);
+                        try
+                        {
+                            if (Enum.TryFormatUnconstrained(value, array, out charsWritten, format))
+                            {
+                                return thisRef.AppendFormatted(array.AsSpan(0, charsWritten));
+                            }
+                        }
+                        finally
+                        {
+                            ArrayPool<char>.Shared.Return(array);
```

Comment:
Isn't the guidance to avoid returning pooled arrays from `finally` blocks, for security reasons? Is the assumption here that the code in the `try` would never throw anyway, so the `try/finally` is just for the convenience to avoid awkward `if`-s/`goto`-s to achieve the same semantics in both branches? 🤔

Reply:
> Isn't the guidance to avoid returning pooled arrays from finally blocks, for security reasons?

No.

> the try/finally is just for the convenience

Yes.

Reply:
> "No."

Not trying to nitpick here, just trying to learn. My understanding was, from several previous conversations with @GrabYourPitchforks on the topic, that "if you're in a `finally` block, you cannot make any assumptions on the state of your program, therefore you should not return rented arrays from a pool there" (I'm paraphrasing). Several other folks and I have since started avoiding this pattern for this reason. There seem to be conflicting messages here, should we come up with one "official" guidance on the topic and maybe integrate that in some XML remarks on the `ArrayPool<T>` docs?

Reply:
In general it's not worth forcibly trying to return arrays to the pool in the case of exceptions because try/finally isn't entirely free and it can lead to more cumbersome code, so the answer to the question "do I need to return arrays in finally blocks" is no.  That different from saying you shouldn't return arrays in finallys, it's just saying you don't have to.  If a try/finally helps with maintainability, it's fine to use.

With regards to reliability, there are certain patterns where things could go badly if you're not careful, e.g.
```C#
char[] array = ArrayPool<char>.Shared.Rent(...);
try
{
    ...
    if (grow)
    {
        ArrayPool<char>.Shared.Return(array);
        array = ArrayPool<char>.Shared.Rent(newSize);
    }
    ...
}
finally
{
    if (array is not null) ArrayPool<char>.Shared.Return(array);
}
```
Here if `Rent` were to throw an exception, we could end up returning the same array to the pool twice, which is why in such patterns you see us instead writing code like:
```C#
char[] toReturn = array;
array = ArrayPool<char>.Shared.Rent(newSize);
ArrayPool<char>.Shared.Return(toReturn);
```
or
```C#
char[] toReturn = array;
array = null;
ArrayPool<char>.Shared.Return(toReturn);
array = ArrayPool<char>.Shared.Rent(newSize);
```
You just need to be cognizant of such patterns.

I've seen some folks express concerns like "well if there's an exception I don't know if some arbitrary code might still be using the array", but that's true regardless of whether there are exceptions or not.

Reply:
Thank you for the very detailed reply Stephen, I appreciate it! 😄

Yup I'm aware that a `try/finally` isn't exactly free (eg. it hurts inlining a lot, AFAIK), and to clarify I didn't mean that one would _have_ to return pooled arrays in a `finally` block, my concern was more about "should I be wary of ever returning pooled arrays from a `finally` block?", which is the part you addressed in the rest of your message.

> "I've seen some folks express concerns like "well if there's an exception I don't know if some arbitrary code might still be using the array", but that's true regardless of whether there are exceptions or not. [...] If there isn't a good reason to use a finally, then on balance, sure, there's less risk to not. But the risk is also low in general, so if there's a reason to use one, eg it makes handling success easier, it's fine to use one."

That's fair, makes sense! Thank you 🙂